flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUpsertKeys.scala

```diff
@@ -186,6 +187,25 @@ class FlinkRelMdUpsertKeys private extends MetadataHandler[UpsertKeys] {
   }
 
   def getUpsertKeys(rel: Window, mq: RelMetadataQuery): JSet[ImmutableBitSet] = {
+    // If it's a ROW_NUMBER Rank, then the upsert keys are partition by key and row number.
+    if (rel.groups.length == 1) {
+      val group = rel.groups.get(0)
+      FlinkRelMdUniqueKeys.INSTANCE.getUniqueKeysOfWindowGroup(group, rel) match {
+        case Some(uniqueKeys) =>
+          val inputKeys = filterKeys(
+            FlinkRelMetadataQuery
+              .reuseOrCreate(mq)
+              .getUpsertKeys(rel.getInput),
+            group.keys)
```

Comment:
This is duplicated with `getUpsertKeysOnOver`. Since you always call `getUpsertKeysOnOver` to get upsert keys derived from child, you can put this at the beginning of the method, and use the result in various places.